### DISCUSSION TITLE
Using brew update, brew upgrade, and brew cleanup when installing, uninstalling and updating packages

### DISCUSSION THREAD

**Discussion by johnmapeson** [Homebrew/discussions#5146] [Getting started] [answered]
### Output of `brew config`

```shell
HOMEBREW_VERSION: 4.1.21
ORIGIN: https://github.com/Homebrew/brew
HEAD: ad1bb178d4a478840c4524063ac5cd52d8667d25
Last commit: 3 months ago
Core tap JSON: 23 Nov 15:31 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 10
Homebrew Ruby: 2.6.10 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: 10-core 64-bit arm_firestorm_icestorm
Clang: 15.0.0 build 1500
Git: 2.39.3 => /Library/Developer/CommandLineTools/usr/bin/git
Curl: 8.4.0 => /usr/bin/curl
macOS: 14.2.1-arm64
CLT: 15.1.0.0.1.1700200546
Xcode: 15.1
Rosetta 2: false
```

### Output of `brew doctor`

```shell
Your system is ready to brew.
```

### Description of issue

I was told by one guy that it is best to use `brew update`, `brew upgrade` and `brew cleanup` as often as possible. That is, that ...

- ... the most safe and clean way to install a package is
  ```
  brew update
  brew upgrade
  brew install foo
  brew cleanup
  ```

- ... the most safe and clean way to uninstall a package is
  ```
  brew update
  brew upgrade
  brew uninstall --ignore-dependencies foo
  brew cleanup
  ```

- and the most safe and clean way to update a package is
  ```
  brew update
  brew upgrade foo
  brew cleanup
  ```

Is it really so?

**Comment by SMillerDev** [comment, marked as answer]
> Is it really so?

Not really, a couple of notes:
>  it is best to use brew update, brew upgrade and brew cleanup as often as possible.

This, for the most part, is true. Homebrew works best when it is updated often. Personally I run `brew upgrade` once a day or so to make sure I have all the latest updates of software for security and possibly new features.
The `brew cleanup` isn't needed as often but it won't hurt to run to clean up some old files. The `brew update` is already run by `brew upgrade` in modern brew, so that one is redundant.

> ... the most safe and clean way to uninstall a package is

This one passes `--ignore-dependencies` which you probably don't want to do, since it will ignore any warnings about other tools that need the package you're removing.

**Comment by johnmapeson** [comment]
> This one passes `--ignore-dependencies` which you probably don't want to do, since it will ignore any warnings about other tools that need the package you're removing.

My intent is to uninstall a package without uninstalling its dependencies, because its dependencies might be used by some other packages.

> Dependencies: You can prevent Homebrew from uninstalling package dependencies using the `--ignore-dependencies` option. You may want to use this option when uninstalling a package that introduces a lot of dependencies that may be used by other packages. You may want to use this option when removing packages such as Ruby, Python, and Node JS. I personally don't use this option and prefer to have a clean state.
>
> ```
> $ brew uninstall --ignore-dependencies postgresql
> ```
>
> Source: https://www.jdeen.com/blog/uninstalling-packages-installed-with-homebrew

So it seems this option makes the process of uninstallation *more* safe, *not* less. Or maybe I don't really understand something?

**Comment by gromgit** [comment]
> My intent is to uninstall a package without uninstalling its dependencies, because its dependencies might be used by some other packages.

Unless something's radically changed during my long Homebrew hiatus, Homebrew has **never** auto-uninstalled a formula's dependencies. @SMillerDev is talking about breaking installed formulae that depend on _the formula you're currently uninstalling_. Passing `--ignore-dependencies` allows that to happen, which will cause you grief down the line.

**Reply by johnmapeson** [reply]
So in other words you mean that using `--ignore-dependencies` when uninstalling a package is generally a bad idea? (Sorry, English is a second language for me, nor I'm familiar with Homebrew terminology either.)

**Reply by SMillerDev** [reply]
Yes, here is an example of two flows:
- You install `check-jsonschema`, which depends on `python@3.12`
- You run `brew uninstall python@3.12`
- Brew stops you with an error since that would break `check-jsonschema`

Or your case:
- You install `check-jsonschema`, which depends on `python@3.12`
- You run `brew uninstall --ignore-dependencies python@3.12`
- Brew ignores the dependencies of `python@3.12` and uninstalls it breaking `check-jsonschema`

**Reply by johnmapeson** [reply]
@SMillerDev Thanks a lot. So, from the blog post above I was sure that using `--ignore-dependencies` is to **prevent** such cases, but it seems that in fact the things are completely opposite?

**Reply by gromgit** [reply]
Yes. The author of that article is completely wrong in this regard.
